### PR TITLE
Thunder integration with OCP

### PR DESCRIPTION
**Pull request by kunallillaney**
Hey Jeremy,

I have merged the latest branch of thunder and documented my function. In addition, the tests are also fixed. They will not fail if OCP is down.

### CODE REVIEW

**Comment by kunallillaney** [comment]
Fixed the bug on our end. All tests now pass for all servers on our end.

**Comment by freeman-lab** [comment]
@kunallillaney do you mind merging in the latest changes from master? I now think the problem is that the test output with your changes became longer that Travis's maximum length, but some tooling I just did simplifies the test output dramatically, and should fix it if that's the problem.

**Comment by kunallillaney** [comment]
@freeman-lab Done and all tests pass. And the test output is now much better and readable. Once your are done with the series changes we can also work on the seriesloader for OCP. Thanks.

File: python/thunder/rdds/fileio/imagesloader.py [hunk 1]
```diff
@@ -133,6 +133,98 @@ def toArray(idxAndBuf):
...
+        import cStringIO
+        pageStr = zlib.decompress ( imgData[:] )
+        pageObj = cStringIO.StringIO ( pageStr )
+        data = load(pageObj)
+        # Data comes in as 4d numpy array in t,z,y,x order. Swapping axes and       removing the time dimension to give back a 3d numpy array in x,y,z order 
+        data = swapaxes(data[0,:,:,:],0,2)
+
+        return data
+      
+      rdd = self.sc.parallelize (enumerate(urlList), len(urlList)).map(lambda (k, v): (k, read(v)))
```
Comment (freeman-lab) [inline python/thunder/rdds/fileio/imagesloader.py]: remove space after `parallelize`

File: python/thunder/rdds/fileio/imagesloader.py [hunk 2]
```diff
@@ -133,6 +133,98 @@ def toArray(idxAndBuf):
         newDims = tuple(list(dims[:-1]) + [nplanes]) if nplanes else dims
         return Images(readerRdd.flatMap(toArray), nrecords=nrecords, dims=newDims, dtype=dtype)
 
+    def fromOCP (self, dataPath, resolution, serverName='ocp.me', startIdx=None, stopIdx=None, minBound=None, maxBound=None ):
```
Comment (freeman-lab) [inline python/thunder/rdds/fileio/imagesloader.py]: Throughout line lengths should be no longer than 120

File: python/thunder/rdds/fileio/imagesloader.py [hunk 3]
```diff
@@ -133,6 +133,98 @@ def toArray(idxAndBuf):
...
+      """Sets up a new Image object with data to read from OCP
```
Comment (freeman-lab) [inline python/thunder/rdds/fileio/imagesloader.py]: Define OCP

File: python/thunder/rdds/fileio/imagesloader.py [hunk 4]
```diff
@@ -133,6 +133,98 @@ def toArray(idxAndBuf):
...
+      # Given a datapath/bucket Query JSON
+      # Given bounds get a list of URI's
+      import urllib2
+      urlList=[]
+      url = 'http://{}/ocp/ca/{}/info/'.format(serverName,dataPath)
+
+      try:
+        f = urllib2.urlopen ( url )
+      except urllib2.URLError, e:
+        raise Exception ( "Faile URL {}".format(url) )
```
Comment (freeman-lab) [inline python/thunder/rdds/fileio/imagesloader.py]: "Faile" -> "Failed"?

File: python/thunder/rdds/fileio/imagesloader.py [hunk 5]
```diff
@@ -133,6 +133,98 @@ def toArray(idxAndBuf):
...
+
+      import json
+      projInfo = json.loads ( f.read() )
```
Comment (freeman-lab) [inline python/thunder/rdds/fileio/imagesloader.py]: Remove extra spaces

File: python/thunder/rdds/fileio/imagesloader.py [hunk 6]
```diff
@@ -133,6 +133,98 @@ def toArray(idxAndBuf):
...
+      ximageSize, yimageSize = projInfo['dataset']['imagesize']['{}'.format(resolution)]
+      zimageStart, zimageStop = projInfo['dataset']['slicerange']
+      timageStart, timageStop = projInfo['dataset']['timerange']
+      xDim, yDim, zDim = cubeDims = projInfo['dataset']['cube_dimension']['{}'.format(resolution)]
+
+      # Checking if dimensions are within bounds
+      if startIdx == None:
+        startIdx = timageStart
+      elif startIdx < timageStart or startIdx > timageStop:
+        raise Exception ( "startIdx out of bounds {},{}".format(timageStart,timageStop) )
```
Comment (freeman-lab) [inline python/thunder/rdds/fileio/imagesloader.py]: Remove extra space before `)`

File: python/thunder/rdds/fileio/imagesloader.py [hunk 7]
```diff
@@ -133,6 +133,98 @@ def toArray(idxAndBuf):
...
+        data = load(pageObj)
+        # Data comes in as 4d numpy array in t,z,y,x order. Swapping axes and       removing the time dimension to give back a 3d numpy array in x,y,z order 
+        data = swapaxes(data[0,:,:,:],0,2)
+
+        return data
+      
+      rdd = self.sc.parallelize (enumerate(urlList), len(urlList)).map(lambda (k, v): (k, read(v)))
+      return Images(rdd, nrecords=len(urlList) )
+
+
```
Comment (freeman-lab) [inline python/thunder/rdds/fileio/imagesloader.py]: Remove extra blank line

File: python/thunder/rdds/fileio/imagesloader.py [hunk 8]
```diff
@@ -133,6 +133,98 @@ def toArray(idxAndBuf):
...
+        imgData = f.read()
+        
+        import zlib
+        import cStringIO
+        pageStr = zlib.decompress ( imgData[:] )
+        pageObj = cStringIO.StringIO ( pageStr )
+        data = load(pageObj)
+        # Data comes in as 4d numpy array in t,z,y,x order. Swapping axes and       removing the time dimension to give back a 3d numpy array in x,y,z order 
+        data = swapaxes(data[0,:,:,:],0,2)
```
Comment (freeman-lab) [inline python/thunder/rdds/fileio/imagesloader.py]: Reformat `data = swapaxes(data[0,:,:,:],0,2)` -> `data = swapaxes(data[0, :, :, :], 0, 2)`

File: python/thunder/rdds/fileio/imagesloader.py (hunk 2 above)
Comment (freeman-lab) [inline python/thunder/rdds/fileio/imagesloader.py]: There are quite a few deviations from PEP conventions for spacing throughout here, can you please open this in PyCharm and try to fix them? That will be more efficient than my marking all of them.

**Comment by freeman-lab** [comment]
@kunallillaney Thanks for the updates. I left a few comments. There are many deviations from PEP conventions for spacing here, can you open in PyCharm and try to fix them?

Also, after looking at it more, I would rather move the top-level `ThunderContext` method outside of `loadImages` and into a separate function called `loadImagesOCP`. The reason is that there is actually little overlap in the input arguments, there are already a large number of arguments for `loadImages`, and this is really a very different format than all the other more standard image formats than the ones currently handled by `loadImages`. Note that this only affects the `loadImages` method, not the `imagesloader.fromOCP` which can remain as is.

Can you try making that change?

**Comment by kunallillaney** [comment]
@freeman-lab Did fix all those(and more) PEP errors in PyCharm. Moved OCP out in the context file. Let me know all this looks good.

**Comment by kunallillaney** [comment]
@freeman-lab Is there any more TODO's on our end?

File: python/thunder/rdds/fileio/imagesloader.py [hunk 9]
```diff
@@ -133,6 +133,107 @@ def toArray(idxAndBuf):
         newDims = tuple(list(dims[:-1]) + [nplanes]) if nplanes else dims
         return Images(readerRdd.flatMap(toArray), nrecords=nrecords, dims=newDims, dtype=dtype)
 
+    def fromOCP (self, bucketName, resolution, serverName='ocp.me', startIdx=None, stopIdx=None, minBound=None,
+                 maxBound=None):
+        """Sets up a new Image object with data to read from OCP
```
Comment (freeman-lab) [inline python/thunder/rdds/fileio/imagesloader.py]: Add definition of OCP

File: python/thunder/utils/context.py [hunk 10]
```diff
@@ -94,12 +94,10 @@ def loadSeries(self, dataPath, nkeys=None, nvalues=None, inputFormat='binary', m
                                      keyType=keyType, valueType=valueType)
         return data
 
-    def loadImages(self, dataPath, dims=None, inputFormat='stack', ext=None, dtype='int16',
-                   startIdx=None, stopIdx=None, recursive=False, nplanes=None, npartitions=None,
-                   renumber=False):
-        """
-        Loads an Images object from data stored as a binary image stack, tif, or png files.
```
Comment (freeman-lab) [inline python/thunder/utils/context.py]: Why did the first line get deleted here?

File: python/thunder/utils/context.py [hunk 11]
```diff
@@ -195,14 +197,63 @@ def loadImages(self, dataPath, dims=None, inputFormat='stack', ext=None, dtype='
...
 
         if not renumber:
             return data
         else:
             return data.renumber()
 
+    def loadImagesOCP(self, bucketName, resolution, serverName='ocp.me', startIdx=None, stopIdx=None,
+                      minBound=None, maxBound=None):
+        """
+        Load Images from OCP
```
Comment (freeman-lab) [inline python/thunder/utils/context.py]: Include definition of OCP

File: python/thunder/utils/context.py [hunk 12]
```diff
@@ -195,14 +197,63 @@ def loadImages(self, dataPath, dims=None, inputFormat='stack', ext=None, dtype='
...
+            These parameters give the starting time (inclusive) and final time (exclusive) of the data files to
+            be read from OCP. For example, startIdx=None (the default) and stopIdx=10 will cause only the first
+            10 data files in dataPath to be read n; startIdx=2 and stopIdx=3 will cause only the third file
+            (zero-based index of 2) to be read in.
+
+        stopIdx: nonnegative int, optional
+            See startIdx.
+        
+        minBound, maxBound: tuple of nonnegative int. optional.
+            X,Y,Z bounds of the data you want to fetch from OCP. minBound contains   the (xMin,yMin,zMin) while
```
Comment (freeman-lab) [inline python/thunder/utils/context.py]: Extra space in `contains    the`

**Comment by freeman-lab** [comment]
Just left a couple final requests, and can you merge in from master to figure out the merge conflict? After that we can merge this in! Thanks.

**Comment by kunallillaney** [comment]
@freeman-lab Done and pushed.

**Comment by freeman-lab** [comment]
Great! LGTM, thanks for the contribution @kunallillaney !

**Comment by kunallillaney** [comment]
Thanks.

**Pull request merged**